csrc/python_frontend/translation.cpp

```diff
@@ -0,0 +1,524 @@
+        }
+
+        // short-circuit: input definition is already visited.
+        if (visited.count(input->definition()) > 0) {
+          continue;
+        }
+
+        to_visit.push_back(input->definition());
+      }
+    }
+    return expression_chain;
+  }
+
+  // Gather the scalar expressions necessary to create the logical domain for a
+  // TensorView.
+  std::vector<Expr*> gatherScalarExpressions(TensorView* tv) {
+    NVF_ERROR(tv != nullptr);
+    std::vector<Expr*> logical_domain_expressions;
+    const std::vector<IterDomain*>& logical_out_domain =
+        tv->domain()->logical();
+    for (IterDomain* id : logical_out_domain) {
+      std::vector<Expr*> extent_definitions =
+          gatherScalarExpressions(id->getMaybeExpandedExtent());
+      logical_domain_expressions.insert(
+          logical_domain_expressions.end(),
+          extent_definitions.begin(),
+          extent_definitions.end());
+    }
+    return logical_domain_expressions;
+  }
+
+  // Check that all of the expression's inputs are defined in FusionDefinition.
+  bool checkExpressionDependencies(Expr* e) {
+    bool check_view_dependency =
+        !e->isA<ViewOp>() || checkViewShapeDependency(e->as<ViewOp>());
+    return check_view_dependency &&
+        std::all_of(e->inputs().begin(), e->inputs().end(), [&](const Val* v) {
+             return map_val_to_fd_index_.count(v) > 0;
+           });
+  }
+
+  void translate() {
+    fd_->setupDefinition();
+
+    // Add Fusion inputs to FusionDefinition
+    for (nvfuser::Val* v : fusion_->inputs()) {
+      OptOutConstDispatch::dispatch(v);
+    }
+
+    // Gather all expressions in CPP Fusion.
+    const std::vector<nvfuser::Expr*> fusion_exprs = fusion_->exprs();
+    std::deque<nvfuser::Expr*> to_visit(
+        fusion_exprs.begin(), fusion_exprs.end());
+
+    // Scalar expressions are not handled by usedMathVals, so gather them
+    // manually.
+    for (Expr* e : to_visit) {
+      if (e->isA<ViewOp>() || e->isA<ExpandOp>() || e->isA<FullOp>()) {
+        std::vector<Expr*> extent_definitions =
+            gatherScalarExpressions(e->output(0)->as<TensorView>());
+        to_visit.insert(
+            to_visit.end(),
+            extent_definitions.begin(),
+            extent_definitions.end());
+      }
+    }
+
+    // Topological search of Fusion expressions
+    size_t skip_count = 0;
+    std::unordered_set<nvfuser::Expr*> visited;
+    while (!to_visit.empty()) {
+      Expr* e = to_visit.front();
+      to_visit.pop_front();
+
+      NVF_ERROR(
+          skip_count <= to_visit.size(),
+          "Cycle detected: None of the expressions can be processed!");
+
+      // short-circuit: skip if already visited
+      if (visited.count(e) > 0) {
+        continue;
+      }
+
+      // short-circuit: skip Split and Merge expressions created by Reshape
+      // short-circuit: skip Resize expressions created by Slice
+      if (e->isA<Split>() || e->isA<Merge>() || e->isA<Resize>()) {
```

Comment:
Do we already check somewhere that no TVs have loop domains, so that it's safe to only check these definitional transforms i.e. that we can skip `Swizzle`?

Reply:
I added a check to compare the loop domain against the logical domain.

https://github.com/NVIDIA/Fuser/pull/2841/files#diff-22914f1fdfa0d5e8983f85f18b2232e2eeadb6df2c28fdb4e7c6ee113d991328R202-R209